File: go/workflows/order_workflow.go
package workflows

import (
	"temporal-order-management/activities"
	"temporal-order-management/app"
	"temporal-order-management/messages"
	"time"

	"github.com/google/uuid"
	"go.temporal.io/sdk/temporal"
	"go.temporal.io/sdk/workflow"
)

func OrderWorkflow(ctx workflow.Context, input app.OrderInput) (output *app.OrderOutput, err error) {
	name := workflow.GetInfo(ctx).WorkflowType.Name
	logger := workflow.GetLogger(ctx)
	logger.Info("Processing order started", "orderId", input.OrderId)

	activityOptions := workflow.ActivityOptions{
		StartToCloseTimeout: 5 * time.Second,
		RetryPolicy: &temporal.RetryPolicy{
			InitialInterval:    1 * time.Second,
			BackoffCoefficient: 2.0,
			MaximumInterval:    30 * time.Second,
		},
	}
	ctx = workflow.WithActivityOptions(ctx, activityOptions)

	localActivityOptions := workflow.LocalActivityOptions{
		StartToCloseTimeout: 5 * time.Second,
	}
	laCtx := workflow.WithLocalActivityOptions(ctx, localActivityOptions)

	// Expose progress as query
	progress, err := messages.SetQueryHandlerForProgress(ctx)
	if err != nil {
		return nil, err
	}

	// Get items
	items := app.Items{}
	err = workflow.ExecuteLocalActivity(laCtx, activities.GetItems).Get(ctx, &items)
	if err != nil {
		return nil, err
	}

	// Check fraud
	err = workflow.ExecuteActivity(ctx, activities.CheckFraud, input).Get(ctx, nil)
	if err != nil {
		return nil, err
	}

	sleep(ctx, 1, progress, 25)

	// Prepare shipment
	err = workflow.ExecuteActivity(ctx, activities.PrepareShipment, input).Get(ctx, nil)
	if err != nil {
		return nil, err
	}

	sleep(ctx, 1, progress, 50)

	// Charge customer
	err = workflow.ExecuteActivity(ctx, activities.ChargeCustomer, input, name).Get(ctx, nil)
	if err != nil {
		return nil, err
	}

	sleep(ctx, 3, progress, 75)

	// Ship order items
	var shipFutures []workflow.Future
	for _, item := range items {
		logger.Info("Shipping item " + item.Description)
		f := workflow.ExecuteActivity(ctx, activities.ShipOrder, input, item)
		shipFutures = append(shipFutures, f)
	}

	// Wait for all items to ship
	for _, f := range shipFutures {
		err = f.Get(ctx, nil)
		if err != nil {
			return nil, err
		}
	}

	sleep(ctx, 1, progress, 100)

	// Generate trackingId
	trackingId := uuid.New().String()
	output = &app.OrderOutput{
		TrackingId: trackingId,
		Address:    input.Address,
	}

	return output, nil
}

func sleep(ctx workflow.Context, seconds int, progress *int, value int) {
	*progress = value
	if seconds > 0 {
		duration := time.Duration(seconds) * time.Second
		workflow.Sleep(ctx, duration)
	}
}

File: ruby/workflows/order_workflow.rb
require 'temporalio/workflow'
require_relative '../models/order_input'
require_relative '../models/order_output'
require_relative '../models/update_order_input'

module Workflows
  class OrderWorkflow < Temporalio::Workflow::Definition
    workflow_name 'OrderWorkflowHappyPath'
    attr_accessor :progress, :retry_policy

    def initialize
      @progress = 0
      @retry_policy = nil
    end

    def execute(input)
      input = Models::OrderInput.new(input['OrderId'], input['Address']) if input.is_a?(Hash)
      workflow_type = Temporalio::Workflow.info.workflow_type
      logger.info("Order workflow started, type = #{workflow_type}, orderId = #{input.order_id}")

      order_items = Temporalio::Workflow.execute_activity(
        Activities::GetItemsActivity,
        start_to_close_timeout: 5
      )

      Temporalio::Workflow.execute_activity(
        Activities::CheckFraudActivity,
        input,
        start_to_close_timeout: 5,
        retry_policy: @retry_policy
      )
      sleep_fn(1, 25)

      Temporalio::Workflow.execute_activity(
        Activities::PrepareShipmentActivity,
        input,
        start_to_close_timeout: 5,
        retry_policy: @retry_policy
      )
      sleep_fn(1, 50)

      Temporalio::Workflow.execute_activity(
        Activities::ChargeCustomerActivity,
        input, workflow_type,
        start_to_close_timeout: 5,
        retry_policy: @retry_policy
      )
      sleep_fn(3, 75)

      activity_handles = order_items.map do |item|
        logger.info("Shipping item: #{item.description}")
        Temporalio::Workflow::Future.new do
          Temporalio::Workflow.execute_activity(
            Activities::ShipOrderActivity,
            input, item,
            start_to_close_timeout: 5,
            retry_policy: @retry_policy
          )
        end
      end

      # Wait for all futures to complete
      results = activity_handles.map(&:wait)

      sleep_fn(0, 100)

      tracking_id = Temporalio::Workflow.random.uuid
      Models::OrderOutput.new(tracking_id, input.address).deep_camelize_keys
    end

    workflow_query(name: 'getProgress')
    def query_progress
      @progress
    end

    private

    def sleep_fn(seconds, progress)
      sleep(seconds) if seconds > 0
      @progress = progress
    end

    def logger
      @logger ||= Temporalio::Workflow.logger
    end
  end
end
